File: app/src/main/java/com/tomclaw/appsend/main/store/search/SearchActivity.java
package com.tomclaw.appsend.main.store.search;

import static com.tomclaw.appsend.util.ThemesKt.updateTheme;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.tomclaw.appsend.R;
import com.tomclaw.appsend.di.legacy.LegacyInjector;
import com.tomclaw.appsend.util.ThemesKt;

@SuppressLint("Registered")
public class SearchActivity extends AppCompatActivity {

    Toolbar toolbar;

    EditText queryEdit;

    SearchFragment searchFragment;

    LegacyInjector legacyInjector;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        updateTheme(this);
        super.onCreate(savedInstanceState);
        if (savedInstanceState == null) {
            legacyInjector.analytics.trackEvent("open-search-screen");
        }
    }

    void init() {
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle(getString(R.string.search_app));
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowTitleEnabled(true);
        }

        queryEdit.setText(searchFragment.query);
        queryEdit.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (searchFragment != null) {
                    searchFragment.filter(s.toString());
                }
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });
    }

    @Override
    protected void onPostResume() {
        super.onPostResume();
        queryEdit.requestFocus();
    }

    boolean actionHome() {
        onBackPressed();
        return true;
    }

    public static Intent createSearchActivityIntent(Context context) {
        return SearchActivity_.intent(context).get();
    }

}

File: app/src/main/java/com/tomclaw/appsend/screen/search/SearchActivity.kt
package com.tomclaw.appsend.screen.search

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.tomclaw.appsend.util.adapter.ItemBinder
import com.tomclaw.appsend.util.adapter.AdapterPresenter
import com.tomclaw.appsend.util.adapter.SimpleRecyclerAdapter
import com.tomclaw.appsend.appComponent
import com.tomclaw.appsend.R
import com.tomclaw.appsend.screen.search.di.SearchModule
import com.tomclaw.appsend.util.updateTheme
import javax.inject.Inject

class SearchActivity : AppCompatActivity(), SearchPresenter.SearchRouter {

    @Inject
    lateinit var presenter: SearchPresenter

    @Inject
    lateinit var adapterPresenter: AdapterPresenter

    @Inject
    lateinit var binder: ItemBinder

    override fun onCreate(savedInstanceState: Bundle?) {
        updateTheme()
        super.onCreate(savedInstanceState)

        val presenterState = savedInstanceState?.getBundle(KEY_PRESENTER_STATE)
        appComponent
            .searchComponent(SearchModule(context = this, presenterState))
            .inject(activity = this)

        setContentView(R.layout.activity_search)

        setupToolbar()

        val adapter = SimpleRecyclerAdapter(adapterPresenter, binder)
        val rootView = window.decorView
        val searchView = SearchViewImpl(rootView, adapter)

        presenter.attachView(searchView)
    }

    private fun setupToolbar() {
        val toolbar = findViewById<androidx.appcompat.widget.Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            setDisplayShowHomeEnabled(true)
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowTitleEnabled(true)
            title = getString(R.string.search_app)
        }
    }

    override fun onStart() {
        super.onStart()
        presenter.attachRouter(this)
    }

    override fun onStop() {
        presenter.detachRouter()
        super.onStop()
    }

    override fun onResume() {
        super.onResume()
        // Request focus on query edit text after resume
        findViewById<android.widget.EditText>(R.id.query_edit)?.requestFocus()
    }

    override fun onDestroy() {
        presenter.detachView()
        super.onDestroy()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBundle(KEY_PRESENTER_STATE, presenter.saveState())
    }

    override fun onOptionsItemSelected(item: android.view.MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressedDispatcher.onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun openAppScreen(appId: String, title: String) {
        val intent = com.tomclaw.appsend.screen.details.createDetailsActivityIntent(
            context = this,
            appId = appId,
            label = title,
            moderation = false,
            finishOnly = true
        )
        startActivity(intent)
    }

}

fun createSearchActivityIntent(context: Context): Intent =
    Intent(context, SearchActivity::class.java)

private const val KEY_PRESENTER_STATE = "presenter_state"

